File: services/message-sender/api/v0/send.go
package sendv0

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/go-playground/validator/v10"
	"github.com/sendgrid/sendgrid-go"
	"github.com/sendgrid/sendgrid-go/helpers/mail"
)

type RequestBody struct {
	Email   string `validate:"required,email"`
	Name    string `validate:"required,min=2"`
	Subject string `validate:"required,min=4"`
	Message string `validate:"required,min=4"`
}

var Validator = validator.New()

func Handler(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		w.WriteHeader(404)
		w.Write([]byte("Not Found!"))
		return
	}

	sgApiKey := os.Getenv("SENDGRID_API_KEY")

	if sgApiKey == "" {
		log.Panicln("Error: Please provide env.SENDGRID_API_KEY")

		payload := map[string]string{
			"error":   "Internal Server Error",
			"message": "Some environment variables is not provided.",
		}

		payloadJson, _ := json.Marshal(payload)

		w.WriteHeader(500)
		w.Header().Set("Content-Type", "application/json")
		w.Write(payloadJson)
		return
	}

	body := new(RequestBody)
	err := json.NewDecoder(r.Body).Decode(&body)

	if err != nil {
		log.Panicln(err)

		payload := map[string]string{
			"error":   err.Error(),
			"message": "Failed to parse request to JSON",
		}

		payloadJson, _ := json.Marshal(payload)

		w.WriteHeader(400)
		w.Header().Set("Content-Type", "application/json")
		w.Write(payloadJson)
		return
	}

	err = Validator.Struct(body)

	if err != nil {
		log.Panicln(err)

		payload := map[string]string{
			"error":   err.Error(),
			"message": "Failed to validate request",
		}

		payloadJson, _ := json.Marshal(payload)

		w.WriteHeader(400)
		w.Header().Set("Content-Type", "application/json")
		w.Write(payloadJson)
		return
	}

	from := mail.NewEmail(body.Name, "test@example.com")
	to := mail.NewEmail("Nurliman Diara Aria", "nurlimandiara@gmail.com")
	replyTo := mail.NewEmail(body.Name, body.Email)
	htmlContent := fmt.Sprintf("<p>%s</p>", body.Message)
	message := mail.NewSingleEmail(from, body.Subject, to, body.Message, htmlContent).SetReplyTo(replyTo)
	client := sendgrid.NewSendClient(sgApiKey)
	response, err := client.Send(message)

	if err != nil {
		log.Panicln(err)

		payload := map[string]string{
			"error":   err.Error(),
			"message": "Error while sending email",
		}

		payloadJson, _ := json.Marshal(payload)

		w.WriteHeader(500)
		w.Header().Set("Content-Type", "application/json")
		w.Write(payloadJson)
		return
	}

	w.WriteHeader(response.StatusCode)
	w.Write([]byte(response.Body))
}

File: services/message-sender/src/handlers/v0/send.rs
use crate::cors::DEFAULT_CORS;
use crate::sendgrid_client::{EmailRecipientSender, SendgridClient};
use serde::Deserialize;
use serde_json::json;
use validator::Validate;
use worker::{console_log, Request, Response, Result, RouteContext};

#[derive(Deserialize, Validate)]
struct RequestBody {
    #[validate(email)]
    email: String,
    #[validate(length(min = 2))]
    name: String,
    #[validate(length(min = 4))]
    subject: String,
    #[validate(length(min = 4))]
    message: String,
}

pub async fn send_message(mut req: Request, ctx: RouteContext<()>) -> Result<Response> {
    let sendgrid_api_key = match ctx.secret("SENDGRID_API_KEY") {
        Ok(value) => value.to_string(),
        Err(_) => String::new(),
    };

    // check SENDGRID_API_KEY
    if sendgrid_api_key.is_empty() {
        console_log!("Error: Please provide env.SENDGRID_API_KEY");

        // init error response body
        let res = Response::from_json(&json!({
            "error": "Internal Server Error",
            "message": "Some environment variables is not provided.",
            "path": req.path(),
        }))?
        .with_cors(&DEFAULT_CORS)?
        .with_status(500);

        // send error response
        return Ok(res);
    }

    let mail_sender_email = match ctx.secret("MAIL_SENDER_EMAIL") {
        Ok(value) => value.to_string(),
        Err(_) => String::new(),
    };

    // check SENDGRID_API_KEY
    if mail_sender_email.is_empty() {
        console_log!("Error: Please provide env.MAIL_SENDER_EMAIL");

        // init error response body
        let res = Response::from_json(&json!({
            "error": "Internal Server Error",
            "message": "Some environment variables is not provided.",
            "path": req.path(),
        }))?
        .with_cors(&DEFAULT_CORS)?
        .with_status(500);

        // send error response
        return Ok(res);
    }

    // parse request body
    let req_body: RequestBody = match req.json::<RequestBody>().await {
        Ok(res) => res,
        Err(error) => {
            let res = Response::from_json(&json!({
                "error": error.to_string(),
                "message": "Failed to parse request to JSON",
                "path": req.path(),
            }))?
            .with_cors(&DEFAULT_CORS)?
            .with_status(400);

            return Ok(res);
        }
    };

    // validate request body
    if let Err(error) = req_body.validate() {
        let res = Response::from_json(&json!({
            "error": error.to_string(),
            "message": "Failed to validate request",
            "path": req.path(),
        }))?
        .with_cors(&DEFAULT_CORS)?
        .with_status(400);

        return Ok(res);
    }

    let sendgrid_client = SendgridClient::new(&sendgrid_api_key);

    // send email
    let sendgrid_res = sendgrid_client
        .send_email(
            EmailRecipientSender {
                // to
                email: ctx.var("MESSAGE_RECIPIENT_EMAIL")?.to_string(),
                name: ctx.var("MESSAGE_RECIPIENT_NAME")?.to_string(),
            },
            EmailRecipientSender {
                // from
                email: mail_sender_email,
                name: "Message Sender Service".to_string(),
            },
            EmailRecipientSender {
                // reply to
                email: req_body.email,
                name: req_body.name,
            },
            &req_body.subject, // subject
            &req_body.message, // message
        )
        .await;

    if let Err(error) = sendgrid_res {
        let res = Response::from_json(&json!({
            "error": error.to_string(),
            "message": "Error while sending email",
            "path": req.path(),
        }))?
        .with_cors(&DEFAULT_CORS)?
        .with_status(500);

        return Ok(res);
    }

    Response::from_json(&json!({ "message": "Message sent!" }))?.with_cors(&DEFAULT_CORS)
}
